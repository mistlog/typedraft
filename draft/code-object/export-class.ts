import { ExportNamedDeclaration, ClassDeclaration, ClassMethod } from "@babel/types";
import { NodePath } from "@babel/core";

export class ExportClassCode {
    m_Path: NodePath<ExportNamedDeclaration>;

    constructor(path: NodePath<ExportNamedDeclaration>) {
        this.m_Path = path;
    }

    get m_Code() {
        return this.m_Path.node;
    }

    get m_Class() {
        return this.m_Code.declaration as ClassDeclaration;
    }

    get m_Members() {
        return this.m_Class.body.body;
    }

    get m_Name() {
        return this.m_Class.id.name as string;
    }

    AddMember(member: ClassMethod) {
        this.m_Members.push(member);
    }
}
