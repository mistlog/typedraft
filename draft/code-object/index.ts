export * from "./export-class";
export * from "./inline-context";
export * from "./local-context";
export * from "./method";
export * from "./module";
