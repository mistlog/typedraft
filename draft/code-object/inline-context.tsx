import { BlockStatement, ExpressionStatement, Statement } from "@babel/types";
import { NodePath } from "@babel/traverse";
import { IDSL } from "../core/transcriber";
import { MatchDSL, __, use } from "draft-dsl-match";

export interface IInlineContext {
    ToStatements: () => Array<Statement>;
}

export class InlineContext {
    m_Path: NodePath<BlockStatement>;

    get m_ID() {
        return Symbol();
    }

    get m_Code() {
        return this.m_Path.node as BlockStatement;
    }

    constructor(path: NodePath<BlockStatement>) {
        this.m_Path = path;
    }

    ToStatements() {
        return this.m_Code.body;
    }

    Resolve(dsl: IDSL) {
        if (dsl.m_Merge) {
            this.m_Path.replaceWithMultiple(dsl.Transcribe(this.ToStatements(), this.m_Path));
        } else {
            this.m_Code.body = dsl.Transcribe(this.ToStatements(), this.m_Path);
        }
    }

    GetDSLName() {
        const statement = this.m_Path.node.body[0] as ExpressionStatement;
        return Î›<string>("match")` ${statement} 
            ${{ expression: { type: "StringLiteral", value: use("text") } }} -> ${(_, { text }) => {
            const [use, dsl_name] = text.trim().split(" ");
            return dsl_name;
        }}
            ${__} -> ${""}
        `;
    }
}
