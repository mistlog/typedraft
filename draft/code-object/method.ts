import {
    ExpressionStatement,
    BinaryExpression,
    JSXElement,
    FunctionExpression,
    JSXIdentifier,
    isIdentifier,
    classMethod,
} from "@babel/types";
import { NodePath } from "@babel/traverse";

/**
 * # Class Method
 * As we want to rearrange code in literate programming, we will seperate class method from class body.
 */
export class MethodCode {
    m_Path: NodePath<ExpressionStatement>;

    constructor(path: NodePath<ExpressionStatement>) {
        this.m_Path = path;
    }

    get m_Code() {
        return this.m_Path.node;
    }

    get m_Left() {
        return this.m_Expression.left as JSXElement;
    }

    get m_Right() {
        return this.m_Expression.right as FunctionExpression;
    }

    get m_Expression() {
        return this.m_Code.expression as BinaryExpression;
    }

    get m_ClassName() {
        return (this.m_Left.openingElement.name as JSXIdentifier).name;
    }

    /**
     * The syntax will be:
     *
    ```ts
    export class Foo {
    }
    
    <Foo/> + function Test(this: Foo, a: number, b: string) {
        return a.toString()+b;
    }
    ```
     *
     * After transcription, we will restore `Foo` class to:
     *
    ```ts
    export class Foo {
        static foo : number;
        Test(a: number, b: string) {
            return a.toString()+b;
        }
    }
    ```
     * 
     * ## Restore: To Class Method
     */
    ToClassMethod() {
        const { id, params: raw_params, body, typeParameters } = this.m_Right;

        /**
         * remove param "this":
         */
        const params = raw_params.filter(param => {
            if (isIdentifier(param) && param.name === "this") {
                return false;
            }

            return true;
        });
        const kind = id.name === "constructor" ? id.name : "method";
        const class_method = classMethod(kind, id, params, body);
        class_method.typeParameters = typeParameters;
        return class_method;
    }
}
