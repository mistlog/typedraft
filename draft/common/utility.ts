import { template } from "@babel/core";
import { Node } from "@babel/types";
import { NodePath } from "@babel/traverse";
import { default as Binding } from "@babel/traverse/lib/scope/binding";
import { parse } from "@babel/parser";
import generate, { GeneratorOptions } from "@babel/generator";

export function ToAst<T = any>(code: string) {
    return (template.ast(code, { plugins: ["typescript", "jsx"] }) as unknown) as T;
}

export function ToNodePath<T = any>(code: string): NodePath<T> {
    const path = new NodePath<T>(null, null);
    path.node = (ToAst(code) as unknown) as T;
    return path;
}

/**
 * only used in test to construct required binding param
 */
export function ToBinding(code: string): Binding {
    const binding = new Binding({
        path: ToNodePath<Node>(code),
        existing: null,
        identifier: null,
        scope: null,
        kind: "let",
    });
    return binding;
}

export function ToFile(raw: string) {
    // babel bug: https://github.com/babel/babel/issues/8837
    // refert to test case "transcriber.interface-no-parse-error"
    const code = raw.replace(new RegExp("^<", "gm"), ";<");
    return parse(code, {
        sourceType: "module",
        plugins: ["typescript", "jsx", ["decorators", { decoratorsBeforeExport: true }]],
    });
}

export function ToString(node: Node, options?: GeneratorOptions) {
    return generate(node, options || {}).code;
}
