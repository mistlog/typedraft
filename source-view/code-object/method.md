# Class Method

As we want to rearrange code in literate programming, we will seperate class method from class body.

```typescript
export class MethodCode {
    m_Code: ExpressionStatement;
    m_Path: NodePath<ExpressionStatement>;
    get m_Left() {
        return this.m_Expression.left as JSXElement;
    }
    get m_Right() {
        return this.m_Expression.right as FunctionExpression;
    }
    get m_Expression() {
        return this.m_Code.expression as BinaryExpression;
    }
    get m_ClassName() {
        return (this.m_Left.openingElement.name as JSXIdentifier).name;
    }
}
```

The syntax would be:

```typescript
export class Foo {
    static foo : number;
}

<Foo/> + function Test(this: Foo, a: number, b: string){
    return a.toString()+b;
}
```

After transcribe, we will restore ```Foo``` class as:

```typescript
export class Foo {
    static foo : number;
    Test(a: number, b: string){
        return a.toString()+b;
    }
}
```

## How to Restore

As described, "\<Foo/> + function Test" is an ```ExpressionStatement```, and we want to extrac infomation in order to restore class method.

```typescript
<MethodCode /> +
    function ToClassMethod(this: MethodCode): IClassMethod {
        const { id, params: raw_params, body } = this.m_Right;

        // remove "this" param
        const params = raw_params.filter(param => isIdentifier(param) && param.name !== "this");
        const kind = id.name === "constructor" ? "constructor" : "method";
        const class_method = { class_name: this.m_ClassName, method: classMethod(kind, id, params, body) };
        return class_method;
    };
```

We can get the "restored" class method now, but we have to find a way to insert it into class. See export-class.

# Trivial

```typescript
export interface IClassMethod {
    class_name: string;
    method: ClassMethod;
}
```

```typescript
<MethodCode /> +
    function constructor(this: MethodCode, node: ExpressionStatement, path?: NodePath<ExpressionStatement>) {
        this.m_Code = node;
        this.m_Path = path;
    };
```