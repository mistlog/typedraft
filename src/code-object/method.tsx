import { ExpressionStatement, BinaryExpression, JSXElement, FunctionExpression, JSXIdentifier, isIdentifier, classMethod, ClassMethod } from "@babel/types";

export interface IClassMethod
{
    class_name: string;
    method: ClassMethod;
}

export class MethodCode
{
    m_Code: ExpressionStatement;

    get m_Left() { return this.m_Expression.left as JSXElement; }
    get m_Right() { return this.m_Expression.right as FunctionExpression; }
    get m_Expression() { return this.m_Code.expression as BinaryExpression; }
    get m_ClassName () { return (this.m_Left.openingElement.name as JSXIdentifier).name; }
}

<MethodCode /> + function constructor(node: ExpressionStatement)
{
    this.m_Code = node;
};

<MethodCode /> + function ToClassMethod(this: MethodCode): IClassMethod
{
    const { id, params: raw_params, body } = this.m_Right;
    const params = raw_params.filter(param => isIdentifier(param) && param.name !== "this");
    const kind = id.name === "constructor" ? "constructor" : "method";

    const class_method = {
        class_name: this.m_ClassName,
        method: classMethod(kind, id, params, body)
    }

    return class_method;
}