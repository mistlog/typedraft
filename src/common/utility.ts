import { template } from "@babel/core";
import { Node } from "@babel/types";
import generate, { GeneratorOptions } from "@babel/generator";
import { parse } from "@babel/parser";

export function ToAst(code: string)
{
    return template.ast(code, { plugins: ["typescript", "jsx"] });
}

export function ToFile(raw: string)
{
    // babel bug: https://github.com/babel/babel/issues/8837
    // refert to test case "transcriber.interface-no-parse-error"
    const code = raw.replace(new RegExp("^<","gm"),";<");
    return parse(code, { sourceType: "module", plugins: ["typescript", "jsx", "optionalChaining"] });
}

export function ToString(node: Node, options?: GeneratorOptions)
{
    return generate(node, options || {}).code;
}