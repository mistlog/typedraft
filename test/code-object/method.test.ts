import { MethodCode } from "../../src/code-object/method";
import { ToAst, ToString } from "../../src/common/utility";
import { ExpressionStatement } from "@babel/types";

describe("method", () => {
    test("method.to-class-method", () => {
        const code = `
            <Foo/> + function Test(this: Foo, a: number, b: string){
                return a.toString()+b;
            }
        `;

        const method = new MethodCode(ToAst(code) as ExpressionStatement);
        const class_method = method.ToClassMethod();

        expect(ToString(class_method)).toMatchSnapshot();
        expect(method.m_ClassName).toEqual("Foo");
    });
});
