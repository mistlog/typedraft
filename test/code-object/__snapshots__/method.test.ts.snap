// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`method to class method 1`] = `
"Test(a: number, b: string) {
  return a.toString() + b;
}"
`;
